### PR TITLE
Provide more X tests for Non-payer Key verification

### PR DESCRIPTION
**Pull request by stoyanov-st**
**Description**:
Additional negative X test scenarios for non-payer keys

**Related issue(s)**:

Fixes #
#9262 #9306 #9307 #9347 #9348 #9302 #9303 #9327 #9328 
**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java [hunk 1]
```diff
@@ -70,6 +74,10 @@
  *     <li>Unfreeze {@code ERC20_TOKEN} via {@link FreezeUnfreezeTranslator#UNFREEZE}.</li>
  *     <li>Transfer {@code ERC20_TOKEN} from  SENDER to RECEIVER.  This should now succeed</li>
  *     <li>Freeze {@code ERC721_TOKEN} without provided freeze key via {@link FreezeUnfreezeTranslator#FREEZE}. This should fail with code TOKEN_HAS_NO_FREEZE_KEY.</li>
+ *     <li>Freeze {@code ERC20_TOKEN} without provided freeze key via {@link FreezeUnfreezeTranslator#FREEZE}. This should fail with code TOKEN_HAS_NO_FREEZE_KEY.</li>
+ *     <li>Unfreeze {@code ERC20_TOKEN} without provided freeze key via {@link FreezeUnfreezeTranslator#FREEZE}. This should fail with code TOKEN_HAS_NO_FREEZE_KEY.</li>
+ *     <li>Unfreeze {@code ERC20_TOKEN} without provided freeze key via {@link FreezeUnfreezeTranslator#FREEZE}. This should fail with code INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE.</li>
```
Comment (lukelee-sl) [inline hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java:79]: This comment and the one below needs to reflect that the given key is the incorrect one.

File: hedera-node/hedera-app/src/xtest/java/contract/GrantRevokeKycXTest.java [hunk 2]
```diff
@@ -72,6 +79,8 @@
  *     <li>Grant KYC {@code ERC721_TOKEN} via {@link GrantRevokeKycTranslator#GRANT_KYC}. This should fail with status INVALID_ACCOUNT_ID.</li>
  *     <li>Grant KYC {@code ERC721_TOKEN} via {@link GrantRevokeKycTranslator#GRANT_KYC}.This should fail with status INVALID_TOKEN_ID.</li>
  *     <li>Transfer {@code ERC721_TOKEN} serial 2345 from SENDER to RECEIVER.  This should now succeed</li>*
+ *     <li>Grant KYC {@code ERC721_TOKEN} via {@link GrantRevokeKycTranslator#GRANT_KYC}.This should fail with status INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE.</li>
```
Comment (lukelee-sl) [inline hedera-node/hedera-app/src/xtest/java/contract/GrantRevokeKycXTest.java:82]: Please verify that these comments match the changes.

**Review by lukelee-sl — DISMISSED** [review]
LGTM

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java [hunk 3]
```diff
@@ -63,9 +74,15 @@
  * <ol>
  *     <li>Wipe {@code ERC721_TOKEN} serial 1234 from Owner's account and verify successful operation</li>*
  *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV1 and verify successful operation</li>*
- *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2 and verify successful operation</li>*
  *     <li>Via {@code assertExpectedAccounts} verify that owner's nft supply was decreased by 1.</li>*
  *     <li>Via {@code assertExpectedTokenRelations} verify that owner's token balance was decreased by 20.</li>*
```
Comment (MrValioBg) [inline hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java:78]: These assertions are done after the Scenario Operations, so we can move the comments to the bottom to keep the chronological order.

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java [hunk 4]
```diff
@@ -158,6 +175,77 @@ protected void doScenarioOperations() {
                         Bytes.wrap(ReturnTypes.encodedRc(INVALID_ACCOUNT_ID).array()),
                         output,
                         "Expected INVALID_ACCOUNT_ID when trying to execute with invalid account address (V2)"));
```
Comment (MrValioBg) [inline hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java:177]: Can we exclude the operations for matching Mono Behaviours in another method like `assertWipeOperationsMatchMonoBehaviour`? Having them with this comment in the middle "// @Future remove to revert #9272 after modularization is completed," can cause confusion and we might mistake that they include the key verification. Hence, we can move them to another method and call it as the last operations.

We can also document it at the top of the class as something like "Via {@code assertWipeOperationsMatchMonoBehaviour} ensure wipe operations in modularization match mono behaviour."

File: hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java [hunk 5]
```diff
@@ -70,6 +74,10 @@
  *     <li>Unfreeze {@code ERC20_TOKEN} via {@link FreezeUnfreezeTranslator#UNFREEZE}.</li>
  *     <li>Transfer {@code ERC20_TOKEN} from  SENDER to RECEIVER.  This should now succeed</li>
  *     <li>Freeze {@code ERC721_TOKEN} without provided freeze key via {@link FreezeUnfreezeTranslator#FREEZE}. This should fail with code TOKEN_HAS_NO_FREEZE_KEY.</li>
```
Comment (MrValioBg) [inline hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java:76]:
```suggestion
 *
```

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java [hunk 6]
```diff
@@ -63,9 +74,15 @@
...
  *     <li>Via {@code assertExpectedAccounts} verify that owner's nft supply was decreased by 1.</li>*
  *     <li>Via {@code assertExpectedTokenRelations} verify that owner's token balance was decreased by 20.</li>*
+ *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2 and verify successful operation</li>*
+ *     <li>Wipe {@code ERC721_TOKEN} serial 1234 from Owner's account. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
+ *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV1. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
+ *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
+ *     <li>Wipe {@code ERC721_TOKEN} serial 1234 from Owner's account. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*
+ *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV1. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*
+ *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*
```
Comment (MrValioBg) [inline hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java:85]: The actual order is:
 *     <li>Wipe {@code ERC721_TOKEN} serial 1234 from Owner's account. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
 *     <li>Wipe {@code ERC721_TOKEN} serial 1234 from Owner's account. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*
 *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV1. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
 *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV1. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*
 *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2. This should fail with INVALID_FULL_PREFIX_SIGNATURE_FOR_PRECOMPILE</li>*
 *     <li>Wipe 10 {@code ERC20_TOKEN} Owner's account via wipeTokenAccountV2. This should fail with TOKEN_HAS_NO_WIPE_KEY</li>*

Comment (MrValioBg) [inline hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java:85]: Those differ a little from the actual order.

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java (hunk 4 above)
Comment (stoyanov-st) [reply hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java:177]: I believe that we will have to change the whole test scenario as now we expect `SUCCESS + Error message` with all the tests in this case, so I don't think we need to externalise the specific "mono behaviour" ones.

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java (hunk 3 above)
Comment (stoyanov-st) [reply hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java]: Done. This was due to merge.

File: hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java (hunk 5 above)
Comment (stoyanov-st) [reply hedera-node/hedera-app/src/xtest/java/contract/FreezeUnfreezeXTest.java]: NIce catch! Thanks.

File: hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java (hunk 6 above)
Comment (stoyanov-st) [reply hedera-node/hedera-app/src/xtest/java/contract/WipeXTest.java]: Rearranged in correct order.

**Review by MrValioBg — APPROVED** [review]
LGTM! 💯

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/RecordFinalizerBase.java [hunk 7]
```diff
@@ -95,7 +95,7 @@ protected Map<EntityIDPair, Long> fungibleChangesFrom(
             final var relTokenId = modifiedRel.tokenId();
             final var token = tokenStore.get(relTokenId);
             // Add this to fungible token transfer list only if this token is a fungible token
-            if (!token.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
+            if (token != null && !token.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/RecordFinalizerBase.java:98]: ⚠️ This hides a critical bug that your tests are surfacing in `FinalizeChildRecordHandler`---the `tokenStore` created [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/FinalizeChildRecordHandler.java#L57) should be the writable store, as apparently(?) the readable version does not include pending modifications to the `tokens` map. (This surprises me also...will keep trying to understand.)

In any case, imo we should revert this change and replace that line in `FinalizeChildRecordHandler` with,
```
         final var tokenStore = context.writableStore(WritableTokenStore.class);
```

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Tests are awesome 🙌 , have a concern about hiding an token `null` case that should be impossible tho. (See commment.)

**Review by tinker-michaelj — APPROVED** [review]
LGTM, ty @stoyanov-st

**Pull request merged**